test/e2e_node/pod_status_test.go

```diff
@@ -0,0 +1,140 @@
+							Delay:    300,
+							ExitCode: 0,
+						}),
+					},
+				},
+			},
+		}
+		preparePod(pod)
+
+		client := e2epod.NewPodClient(f)
+		pod = client.Create(ctx, pod)
+
+		ginkgo.By("Waiting for the pod's status to become Running")
+		err := e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Getting the current pod sandbox ID")
+		rs, _, err := getCRIClient()
+		framework.ExpectNoError(err)
+		sandboxes, err := rs.ListPodSandbox(ctx, &runtimeapi.PodSandboxFilter{
+			LabelSelector: podLabels,
+		})
+		framework.ExpectNoError(err)
+		gomega.Expect(sandboxes).To(gomega.HaveLen(1))
+		podSandboxID := sandboxes[0].Id
+
+		// We need to wait for the pod to be Running before simulating the node reboot,
+		// to avoid any unintended effects from the previous init container state.
+		// Simulate node reboot by restarting the kubelet and the pod sandbox.
+		ginkgo.By("Stopping the kubelet")
+		startKubelet := stopKubelet()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeFalseBecause("kubelet should be stopped"))
+
+		ginkgo.By("Stopping the pod sandbox")
+		err = rs.StopPodSandbox(ctx, podSandboxID)
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Starting the kubelet")
+		startKubelet()
+		kubeletRestartedTime := time.Now().UTC()
+		gomega.Eventually(ctx, func() bool {
+			return kubeletHealthCheck(kubeletHealthCheckURL)
+		}, f.Timeouts.PodStart, f.Timeouts.Poll).Should(gomega.BeTrueBecause("kubelet should be started"))
+
+		ginkgo.By("Waiting for the regular init container to be started after the node reboot")
+		err = e2epod.WaitForPodInitContainerStarted(ctx, f.ClientSet, pod.Namespace, pod.Name, 0, f.Timeouts.PodStart)
+		framework.ExpectNoError(err)
+
+		pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+		gomega.Expect(pod.Status.Phase == v1.PodPending).To(gomega.BeTrueBecause("pod should be pending during the execution of the init container after the node reboot"))
+
+		ginkgo.By("Get the logs of the init container after the kubelet restart")
+		logs, err := f.ClientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &v1.PodLogOptions{
+			Container: init,
+			SinceTime: &metav1.Time{ // Use the time when the kubelet was restarted as the start point for the logs
+				Time: kubeletRestartedTime,
+			},
+		}).DoRaw(ctx)
+		framework.ExpectNoError(err)
+		// The `container Starting` log is generated in ExecCommand.
+		expectedLog := fmt.Sprintf("%s Starting", init)
+		gomega.Expect(string(logs)).To(gomega.ContainSubstring(expectedLog), "The logs should contain the expected output from the init container after the kubelet restart")
+
+		ginkgo.By("Verifying that the pod fully starts")
+		err = e2epod.WaitForPodNameRunningInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace)
+		framework.ExpectNoError(err)
+	})
```

Comment:
It might be a good idea to add a step to verify that the initContainer has successfully completed after the Pod reaches the Running state.

```go
		ginkgo.By("Verifying that the init container completed successfully")
		pod, err = client.Get(ctx, pod.Name, metav1.GetOptions{})
		framework.ExpectNoError(err)
		gomega.Expect(pod.Status.InitContainerStatuses[0].State.Terminated).NotTo(gomega.BeNil(), "init container should be terminated")
		gomega.Expect(pod.Status.InitContainerStatuses[0].State.Terminated.ExitCode).To(gomega.BeZero(), "init container should complete successfully")
```

Reply:
Thank you, this looks great.

Reply:
Thank you for your suggestion, but I don't think it's necessary to do so. A pod running is certainly after the regular init container has completed.

Reply:
Yes, but that's just our assumption, and we might break this behavior in the future. Since the main point of this test is the init container, I personally think it’s worth checking that it completes successfully after the Pod starts.
I don’t think it’d take much effort to add it here, but if you don’t think it’s necessary, I’ll leave it up to you.

Reply:
Thanks